mlflow/sklearn/__init__.py

```diff
@@ -500,3 +507,90 @@ def should_log(self):
         return (self._parent is None) or (
             self._parent.allow_children and self._parent.clazz != self.clazz
         )
+
+
+def _is_old_version():
+    from sklearn import __version__
+
+    return LooseVersion(__version__) < LooseVersion("0.20.3")
+
+
+def autolog():
+    """
+    Enable autologging for scikit-learn.
+    """
+    import sklearn
+
+    if _is_old_version():
+        warnings.warn(
+            "Autologging utilities may not work properly on scikit-learn < 0.20.3 "
+            "(current version: {})".format(sklearn.__version__),
+            stacklevel=2,
+        )
+
+    def fit_mlflow(self, func_name, *args, **kwargs):
+        active_run_exists = mlflow.active_run() is not None
+        if not active_run_exists:
+            try_mlflow_log(mlflow.start_run)
+
+        # TODO: We should not log nested estimator parameters for
+        # parameter search estimators (GridSearchCV, RandomizedSearchCV)
+        try_mlflow_log(mlflow.log_params, self.get_params(deep=True))
+        try_mlflow_log(
+            mlflow.set_tags,
+            {
+                "estimator_name": self.__class__.__name__,
+                "estimator_class": self.__class__.__module__ + "." + self.__class__.__name__,
+            },
+        )
+
+        original_fit = gorilla.get_original_attribute(self, func_name)
+        try:
+            fit_output = original_fit(*args, **kwargs)
+        except Exception as e:
+            _logger.warning("{} failed: {}".format(original_fit.__qualname__, str(e)))
+            mlflow.end_run(RunStatus.to_string(RunStatus.FAILED))
+            return
```

Comment:
Should we raise an exception here?

Reply:
Yep! We'll want to re-raise the scikit exception so the user can observe / handle it. We should add a test case for this behavior too

Reply:
Also, we'll only want to end the run if the autologging routine started it. Otherwise, the following code won't behave as desired:

```
import mlflow.sklearn
mlflow.sklearn.autolog()

with mlflow.start_run():
    try:
        throwing_model.fit(X, y)
    catch Exception:
        # Do something else, run should continue...
```

Reply:
got it! I'll fix the code